### ISSUE TITLE
How to integrate calico network policy engine with aws vpc cni plugin?

### ISSUE THREAD

**Issue by amitkatyal** [aws/amazon-vpc-cni-k8s#1573]
We have a requirement to integrate the calico network policy engine along with the vpc cni plugin to support the k8s network policies and have done the same by installing calico operator and CRS via the k8s manifests (https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/master/config/v1.9/calico-operator.yaml & https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/master/config/v1.9/calico-crs.yaml). However, still having the following queries,

1. aws-calico helm charts are to install the calico CNI which also takes of k8s POD networking as well right? So, if the requirement is to install only the network policy engine, we should do it via the manifest, or is there a way to install the network policy engine via the helm chart?

2.  Ater installing the calico network policy engine, we are not able to upgrade the EKS. We are deploying the EKS via the AWS quick start cloud formation template guide but on upgrading from k8s version 1.20 to 1.21, Managed node group upgrade is failing. We found that on manage node group upgrade when the new nodes are added to the cluster, then the tigera-operator is adding more instances of typha to the cluster. For example, for a two-node cluster, when two more nodes are added, then the tigera-operator is adding two more instances of the typha to the cluster. This is resulting in 4 instances of typha (for a two-nodes cluster). Now, when the upgrade process is trying to take the old nodes out from the cluster, these typha pods are not evicting due to which the upgrade process is failing with a reason "Manage node group upgrades failed due to pod eviction failure).

We tried scaling down the operator during the upgrade process and it seems to be helping but is there any other workaround available that can be applied during the installation time itself so that we don't need to apply any workaround during the upgrade process as automating the workaround during an upgrade process is another pain?

**Comment by globus68** [comment]
We upgraded to VPC-CNI v1.9.0-eksbuild.1 last week, and were happy with the increased pod density. However finding out the addon did not support network policies, we are also wondering how we can use Calico (for network policy support) together with the new plugin to still support increased pod density. If this _is_ still possible, please lead us in the right direction, this is a little bit of "gray matter" for the inexperienced :)

**Comment by amitkatyal** [comment]
@globus68 Did you try installing the Calico network policy engine along with VPC CNI plugin 1.9 as per the AWS documentation?

https://docs.aws.amazon.com/eks/latest/userguide/calico.html

**Comment by globus68** [comment]
@amitkatyal Yes, I have installed it according to the documentation. However, using t3.small instances the pre-v9 version of the plugin gave 11 pods available. After installing the v9 plugin we got 110 available pods, which is very good. But when installing Calico on top of that, it seems we have returned to 11 available pods. I base this on having two instances; after installing Calico we had 22 running pods, the rest in pending state. 22 is max for two instances pre-v9. After adding another instance (3 altogether) the remaining pending pods also started. My assumption then is that installing Calico overrides the v9 VPC-CNI plugin leading to the old count of available pods. Please note that this area of k8s is a bit new and unknown to us...

**Comment by amitkatyal** [comment]
@globus68 Thanks for the detailed reply!

**Comment by amitkatyal** [comment]
@jayanthvn Any suggestion? Thanks!

**Comment by jayanthvn** [comment]
@amitkatyal - regarding the queries -

aws-calico doesn't install Calico CNI, we will still use AWS VPC CNI. You can use the manifest - (https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/master/config/v1.9/calico-operator.yaml & https://raw.githubusercontent.com/aws/amazon-vpc-cni-k8s/master/config/v1.9/calico-crs.yaml) or we do have the helm chart now - https://github.com/aws/amazon-vpc-cni-k8s/tree/master/charts/aws-calico. The helm chart is updated recently and is in sync with calico repo.

Reg the upgrade issue, we will repro and get back to you.

**Comment by jayanthvn** [comment]
@globus68 - Installing calico network policy should not impact number of pods/IPs. So when you say pending pods, was the pod scheduled on the node and IPs weren't allocated? Also on one of the instances, can you please share me the maxPods you have configured?

**Comment by amitkatyal** [comment]
@jayanthvn Thanks a lot for your quick reply!
I Will try out the helm chart and wait for the feedback on the upgrade issue. Thanks.

**Comment by amitkatyal** [comment]
@jayanthvn I tried installing the calico helm charts but the installation failed with the below error. However, the installation works with the manifest.

It is failing due to - Readiness probe failed: calico/node is not ready: Felix is not ready: readiness probe reporting 503
Please note that I'm trying to install the helm charts with default values.

VPC CNI - I am using the latest AWS VPC CNI helm chart version (1.1.8)

![image](https://user-images.githubusercontent.com/42130809/129787632-bc1eff5e-ecf8-4e9a-a4b6-fb5ecdd1a2b9.png)

**Comment by jayanthvn** [comment]
thanks for checking @amitkatyal . Will check why calico node is failing maybe some dependency. Since we generate the manifest from the same charts and store it in master/config.

**Comment by amitkatyal** [comment]
@jayanthvn Thanks, will wait for your findings. Please let me know if you would want me to do any tests.

**Comment by globus68** [comment]
@jayanthvn @amitkatyal Sorry, my bad! I did not do all the steps required in the installation of the plugin documented here: https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html. I will do that and report back. Sorry, again, must improve my documentation reading skills...

**Comment by globus68** [comment]
@jayanthvn @amitkatyal I can confirm everything is working. The updated plugin and Calico (installed using yaml manifests from this Github) works like a charm. I have now currently 17 pods running on a t3.small instance, and I guess I can add even more. Thank you for the good work and sorry about my confusion

**Comment by globus68** [comment]
@jayanthvn @amitkatyal An update: as mentioned, the combination addon v1.9/calico works, e.g having 17 pods on a t3.small instance, so far so good. However, applying AWS Load Balancer Controller v2.2.0 gives me "error: cannot resolve pod ENI for pods" in the load balancer's log when applying one of our applications to the cluster. This may seem to be reported in https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/2116, but due to lack of knowledge 2116 might be another issue. I have tried using the image shoekstra/aws-alb-ingress-controller:latest-patch for the load balancer, but I still get the same error message. Any ideas?

**Comment by amitkatyal** [comment]
@jayanthvn, Any update on the calico helm charts? I think I forgot to mention that when I had deployed the calico helm chart, the tigera operator was not deployed.

**Comment by tscn** [comment]
@amitkatyal We are experiencing the same issue with upgrading two-node EKS-clusters that use tigera-operator to manage Calico. We have reported this here (including our assumption how to solve it): https://github.com/tigera/operator/issues/1441
The helm chart with proportional autoscaler works for us but we are deploying Calico 3.17 at the moment.

**Comment by amitkatyal** [comment]
@tscn Thanks for the update!
For the upgrade issue, will wait for the update from the tigera.
Regarding the helm chart, I will check again and get back.

**Comment by jayanthvn** [comment]
We will look into helm install issue as part of this PR - https://github.com/aws/amazon-vpc-cni-k8s/pull/1578

**Comment by amitkatyal** [comment]
Sorry for the late reply!
Regarding the helm chart, I was getting an error while installing the CRDS via the cloud formation template. However, the manifest file approach works fine with the CF template.

**Comment by amitkatyal** [comment]
@jayanthvn @haouc Kindly let me know when are we planning to merge this PR - #1578 so that I can plan it accordingly. Thanks!

**Comment by jayanthvn** [comment]
@amitkatyal #1578 is merged to master.

**Comment by amitkatyal** [comment]
Thanks, @jayanthvn for the update!

**Comment by jayanthvn** [comment]
Yes, I will track it for 1.10 or for any next CNI release.

**Comment by amitkatyal** [comment]
Thanks a lot, will wait for the 1.10!

**Comment by jayanthvn** [comment]
Will close the issue for now, please feel free to reach out if you need any help.

**Comment by amitkatyal** [comment]
@jayanthvn We are deploying the calico operator manifest using the CloudFormation Registry resource provided by the AWS quick start - https://github.com/aws-quickstart/quickstart-kubernetes-resource-provider/blob/f54ddce8f89cd3996b0ba79f1a0ada1945e8c539/apply/src/awsqs_kubernetes_resource/handlers.py

The stack creation is working fine. However, the update stack failed due to the timeout. After further investigation, found that CF service expects the response within 60 seconds, but since the manifest is around 300 KB, the CloudFormation plugin takes approx 5 seconds to download the manifest, then around 8-10 seconds in checking if the resource already exists, and finally performs the kubectl apply step. All these steps take a little longer than 60 seconds and result in the CF service timeout due to which update fails. 

Just trying to understand how do you recommend the Calico Policy Network Engine installation with the VPC CNI in the production cluster?. 

I was also checking the helm charts calico repo, it also has Kubernetes "customization" dependency before installing the helm charts. Since  we are deploying the stack using the CloudFormation template, we don't have a custom plugin to deploy the resource using "Kubernetes customization" [ kubectl apply -k github.com/aws/eks-charts/tree/master/stable/aws-calico/crds]

https://github.com/aws/eks-charts/tree/master/stable/aws-calico


I was also looking at the Calico Manifest and it seems to be deploying a lot of k8s objects, Do we need all of them for the Calico network policy engine along with the VPC CNI use-case? 

Could you please suggest how to automate the Calico Policy Engine  + VPC CNI using the CloudFormation template, as we are completely blocked on the update issue? Thanks.
